### ISSUE TITLE
Analyzer: Bogus FixedAddressDereference warning for non-constant pointer

### ISSUE THREAD

**Issue by anarazel** [llvm/llvm-project#197524]
Hi,

In the attached, greatly simplified (and pointless-ified), postgres code, FixedAddressDereference triggers a bogus warning.
[policy_edit.c](https://github.com/user-attachments/files/27724522/policy_edit.c)

```
$ clang-23 --analyze -Xclang -analyzer-checker=optin.core.FixedAddressDereference --analyzer-output text -xc policy_edit.c
policy_edit.c:21:10: warning: Array access (via field 'data') results in a dereference of a fixed address
      [optin.core.FixedAddressDereference]
   21 |                 return tup->data[narg];
      |                        ^
policy_edit.c:35:6: note: Assuming 'polcmd' is equal to 17
   35 |         if (polcmd != 17 && stmt->with_check != 0)
      |             ^~~~~~~~~~~~
policy_edit.c:35:19: note: Left side of '&&' is false
   35 |         if (polcmd != 17 && stmt->with_check != 0)
      |                          ^
policy_edit.c:41:17: note: Passing pointer value (Tuple *)17 via 1st parameter 'tup'
   41 |         return get_att(policy_tuple, argbyval, 1);
      |                        ^~~~~~~~~~~~
policy_edit.c:41:9: note: Calling 'get_att'
   41 |         return get_att(policy_tuple, argbyval, 1);
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
policy_edit.c:18:6: note: Assuming the condition is false
   18 |         if (argbyval[narg] == 0)
      |             ^~~~~~~~~~~~~~~~~~~
policy_edit.c:18:2: note: Taking false branch
   18 |         if (argbyval[narg] == 0)
      |         ^
policy_edit.c:21:10: note: Array access (via field 'data') results in a dereference of a fixed address
   21 |                 return tup->data[narg];
      |                        ^    ~~~~
1 warning generated.
```

Note how,  after comparing polcmd with 17, the analyzer thinks the tuple arguments to get_att is 17 as well. Which makes no sense, afaict.

Greetings,

Andres

**Comment by NagyDonat** [comment]
Further simplifications of the code example yields this code, which [still triggers](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:2,endLineNumber:13,positionColumn:2,positionLineNumber:13,selectionStartColumn:2,selectionStartLineNumber:13,startColumn:2,startLineNumber:13),source:'typedef+unsigned+long+int+uintptr_t%3B%0A%0Atypedef+struct+Tuple+%7B%0A%09uintptr_t+data%5B8%5D%3B%0A%7D+Tuple%3B%0A%0Auintptr_t+AlterPolicy(Tuple+*policy_tuple)+%7B%0A%09uintptr_t+polcmd%3B%0A%09polcmd+%3D+(uintptr_t)%26policy_tuple-%3Edata%5B0%5D%3B%0A%09if+(polcmd+!!%3D+17)%0A%09%09return+0%3B%0A%09return+policy_tuple-%3Edata%5B1%5D%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:43.434424933111714,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:clang2110,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!((name:boost,ver:'178'),(name:outcome,ver:trunk)),options:'-Wall+-Wextra+',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+21.1.0+(Editor+%231)',t:'0')),header:(),k:56.28870559521248,l:'4',m:52.159468438538205,n:'0',o:'',s:0,t:'0'),(g:!((h:tool,i:(args:'--checks%3D!'clang-analyzer-optin.core.FixedAddressDereference!'',argsPanelShown:'0',compilerName:'x86-64+clang+21.1.0',editorid:1,fontScale:14,fontUsePx:'0',j:1,monacoEditorHasBeenAutoOpened:'1',monacoEditorOpen:'1',monacoStdin:'1',stdin:'',stdinPanelShown:'1',toolId:clangtidytrunk,treeid:0,wrap:'1'),l:'5',n:'0',o:'clang-tidy+(trunk)+x86-64+clang+21.1.0+(Editor+%231,+Compiler+%231)',t:'0')),header:(),l:'4',m:47.840531561461795,n:'0',o:'',s:0,t:'0')),k:44.11472315495563,l:'3',n:'0',o:'',t:'0'),(g:!((h:output,i:(compilerName:'x86-64+clang+(trunk)',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+21.1.0+(Compiler+%231)',t:'0')),k:12.45085191193264,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) an equivalent report.

```c
typedef unsigned long int uintptr_t;

typedef struct Tuple {
	uintptr_t data[8];
} Tuple;

uintptr_t AlterPolicy(Tuple *policy_tuple) {
	uintptr_t polcmd;
	polcmd = (uintptr_t)&policy_tuple->data[0];
	if (polcmd != 17)
		return 0;
	return policy_tuple->data[1];
}
```
(In the original reproducer the contents of `argbyval` were unknown, so at the `if (argbyval[narg] == 0)` checks the analyzer examined both possibilities. The bug report is generated on the execution path that takes `return (uintptr_t)&tup->data[narg];` in the first `get_att` and `return tup->data[narg];` in the second one -- so I inlined these into my shortened reproducer.  Note that this is a reasonable code path: nothing prevents `argbyval[0] == 0` and `argbyval[1] != 0`.)

Although overall this result is clearly unwanted, I don't see a concrete step where the analyzer "makes a mistake": it follows a plausible execution path, and when it sees the comparison of a symbol and a number, it assumes that this symbol (`(uintptr_t)policy_data`) may be equal to that number (because otherwise why would we compare them?).

Of course when we look at the whole picture we can see that the `polcmd != 17` check is intended for the cases when `argbyval[0] != 0` (i.e. we're comparing the _value_ of element 0 and not the _address_ of element 0) -- but the static analyzer focuses on one execution path and cannot see that a check is more relevant on other branches.

Also, a check can be relevant on the current path even if a check was intended for some other one. In this particular example we can see that a `Tuple *` is aligned to 8 bytes, so it cannot be equal to 17 -- but if 17 was replaced by 0xdedbeef0, then the analyzer should take this equality as plausible.

Overall, I feel that this is an unfortunate but not entirely buggy result from the analyzer.

**Comment by anarazel** [comment]
> Although overall this result is clearly unwanted, I don't see a concrete step where the analyzer "makes a mistake": it follows a plausible execution path, and when it sees the comparison of a symbol and a number, it assumes that this symbol ((uintptr_t)policy_data) may be equal to that number (because otherwise why would we compare them?).

But why is the analyzer concluding that a comparison of policy_tuple->data[0] == 17 implies anything about the value of policy_tuple itself? That's imo pretty confusing / plainly wrong?

Note that in my repro (and in the real case), the branch tests for *two* conditions (`polcmd != 17 && stmt->with_check != 0`). So even if `polcmd != 17` one cannot infer that in the else of that branch (or after a return inside the branch) `polcmd` is 17.


> Overall, I feel that this is an unfortunate but not entirely buggy result from the analyzer.

To me it seems clearly buggy to take the result of a comparison of a struct member and infer from that that the pointer to the struct has that value? What am I missing?

**Comment by NagyDonat** [comment]
> But why is the analyzer concluding that a comparison of policy_tuple->data[0] == 17 implies anything about the value of policy_tuple itself? That's imo pretty confusing / plainly wrong?

You're missing that the code contains an address-of operator (which is admittedly in a bit unusual context).

The comparison is `(uintptr_t)&(policy_tuple->data[0]) == 17`, i.e. instead of comparing _the value stored in_ `policy_tuple->data[0]`, the code is comparing the _address of_  `policy_tuple->data[0]` (as a pointer-sized integer). As `data` is the first field of the struct and `0` is zero, `&(policy_tuple->data[0])` points to the same place as `policy_tuple`.

> Note that in my repro (and in the real case), the branch tests for _two_ conditions (`polcmd != 17 && stmt->with_check != 0`). So even if `polcmd != 17` one cannot infer that in the else of that branch (or after a return inside the branch) `polcmd` is 17.

The presence of the `return` is not relevant for the analyzer, in fact, it doesn't _see_ the `return` statement on the analysis path that leads to reporting this error.

When the analyzer sees an equality or inequality operator (like `polcmd != 17`), it immediately [1] splits the analysis process into two analysis paths: one where the comparison is assumed to be true, and one where the comparison is assumed to be false.
-  If one of these assumptions contradicts the previous knowledge and assumptions of the analyzer, then the corresponding analysis path is rejected (and the analysis continues only on the feasible path).
- If both assumptions are compatible with the previous knowledge and assumptions, then the analyzer will follow both paths independently of each other (until they return from the analysis entrypoint function, or run into a situation that is reported as a bug, or exhaust the budget [number of analysis steps] of the analyzer).

For the analyzer, the presence of `polcmd != 17` is a hint that this is sometimes true and sometimes false -- and the analyzer cannot understand that this check is only relevant in the case when `argbyval[0]` is nonzero (and therefore we will compare the _value_ read from the tuple with the 17 instead of the _address_ of the tuple element).

In general, the analyzer often produces unwanted results in cases when logical connections between conditionals are not visible in the source code. For example, consider the following slightly contrived class:
```c++
class Foo {
  int *Data;
  int Size = 0;
public:
  Foo(int ReserveSize = 0, int FirstElem = 0) {
    if (ReserveSize)
      Data = new int[ReserveSize];
    if (FirstElem)
      Data[Size++] = FirstElem;
  }
};
```
This may be part of a valid program if callers of this constructor always specify nonzero `ReserveSize` when they specify nonzero `FirstElem` -- but if this constructor is analyzed as the entrypoint, the analyzer will consider the case when `ReserveSize == 0` and `FirstElem != 0` and report that this dereferences an uninitialized pointer.

-------

[1] Strictly speaking, this is the eagerly-assume mode of the analyzer, which is enabled by default. However, even if eagerly-assume is disabled, the analyzer needs to separate the analysis paths when it encounters a so-called terminator statement (e.g. conditions of `if`, `while` etc., the LHS of a short-circuiting `&&` or `||`) because it always follows one execution path and cannot be in a "quantum superposition" of entering and not entering the `if` body.
